econml/ortho_forest.py

```diff
@@ -548,28 +630,67 @@ def fit(self, Y, T, X, W=None, inference='auto'):
         -------
         self: an instance of self.
         """
-        return super().fit(Y, T, X, W=W, inference=inference)
+        Y, T, X, W = check_inputs(Y, T, X, W)
+
+        if self.discrete_treatment:
+            d_t_in = T.shape[1:]
+            T = self._one_hot_encoder.fit_transform(T.reshape(-1, 1))
+            self._d_t = T.shape[1:]
+            self.transformer = FunctionTransformer(
+                func=_EncoderWrapper(self._one_hot_encoder).encode,
+                validate=False)
+
+        if self.global_residualization:
+            cv = check_cv(self.global_res_cv, y=T, classifier=self.discrete_treatment)
+            cv = list(cv.split(X=X, y=T))
+            Y = Y - cross_val_predict(self.model_Y_final, self._combine(X, W), Y, cv=cv, safe=False).reshape(Y.shape)
+            T = T - cross_val_predict(self.model_T_final, self._combine(X, W), T, cv=cv, safe=False).reshape(T.shape)
+
+        super().fit(Y, T, X, W=W, inference=inference)
+
+        # weirdness of wrap_fit. We need to store d_t_in. But because wrap_fit decorates the parent
+        # fit, we need to set explicitly d_t_in here after super fit is called.
+        if self.discrete_treatment:
+            self._d_t_in = d_t_in
+        return self
 
     def const_marginal_effect(self, X):
         X = check_array(X)
         # Override to flatten output if T is flat
-        effects = super(ContinuousTreatmentOrthoForest, self).const_marginal_effect(X=X)
-        if not self._d_t:
-            # T is flat
-            return effects.flatten()
-        return effects
+        effects = super().const_marginal_effect(X=X)
+        return effects.reshape((-1,) + self._d_y + self._d_t)
     const_marginal_effect.__doc__ = BaseOrthoForest.const_marginal_effect.__doc__
 
     @staticmethod
-    def nuisance_estimator_generator(model_T, model_Y, random_state=None, second_stage=True):
+    def nuisance_estimator_generator(model_T, model_Y, random_state=None, second_stage=True,
+                                     global_residualization=False, discrete_treatment=False):
```

Comment:
`global_residualization` is only used when calling `_get_conforming_residuals`. Instead of having this parameter in the `nuisance` or `parameter` estimators, you can instead modify `get_conforming_residuals`:

```
@staticmethod
def _get_conforming_residuals(Y, T, nuisance_estimates, global_residualization):
    if nuisance_estimates==0:
        return reshape_Y_T(Y, T)
    # returns shape-conforming residuals
    Y_hat, T_hat = reshape_Y_T(*nuisance_estimates)
    Y, T = reshape_Y_T(Y, T)
    Y_res, T_res = Y - Y_hat, T - T_hat
    return Y_res, T_res
```

Reply:
I had tried to do this but failed to do in a simple manner. The problem is that _get_conforming_residuals is a static function and cannot access the self.globa_residualization. So one would need to write a generator in a similar fashion that generates the right stattic function. But then it's weird how to merge this with init etc. I found the above cleaner, as we already have generators for these other methods.

Reply:
Sorry, in the example above, global_residualization is not necessary since it just checks whether the nuisances are `0`. It's a sort of contract, when residualization is global, the nuisances are 0 and that can be used by `_get_conforming_residuals`

Reply:
what if nuisance estimates just happen to be 0 (weird corner case), but weird to use a numeric value to have a special meaning.

Reply:
We could change the nuisance estimator to return the string 'global' and then check for 'global' here. What do you think?

Reply:
Let me know if you think I should change to something like this. Happy either way.

Reply:
Note that `nuisance_estimates` is usually a tuple of two arrays, so being `0` is a very special case. We can also do `global` or `np.nan`. If none of these seem appealing, we can also just make the method non-static. Either of these will remove a lot of overhead.

Reply:
Sounds good. I'll hcange to global. Though still we need the global_residualization in the generator of the nuisance estimator, as it needs to return the 0 or the global key when global residualziation is enabled. But indeed we are removing the other generators

Reply:
Made changes. Let me know

Reply:
Looks good!